### PR TITLE
Make Adam optimizer differentiable

### PR DESCRIPTION
**Pull request by emcastillo**
Continues [80938](https://github.com/pytorch/pytorch/pull/80938)

### CODE REVIEW

File: torch/optim/adam.py [hunk 1]
```diff
@@ -278,7 +284,10 @@ def _single_tensor_adam(params: List[Tensor],
 
             if amsgrad:
                 # Maintains the maximum of all 2nd moment running avg. till now
-                torch.maximum(max_exp_avg_sqs[i], exp_avg_sq, out=max_exp_avg_sqs[i])
+                if differentiable:
+                    max_exp_avg_sqs[i][:] = torch.maximum(max_exp_avg_sqs[i].clone(), exp_avg_sq)
```
Comment (albanD) [inline torch/optim/adam.py:288]: nit: `max_exp_avg_sqs[i].copy_(torch.maximum(max_exp_avg_sqs[i], exp_avg_sq))`

Also the perf hit compared to the out= call below is really minor. I think we can unconditionally use this new version for the sake of simplicity (same below). cc @jbschlosser what do you think?

**Review by albanD — COMMENTED** [review]
Nice!

File: torch/optim/adam.py (hunk 1 above)
Comment (emcastillo) [reply torch/optim/adam.py]: One thing here is that we need to clone `max_exp_avg_sqs` for this to work ...
Wonder if its ok to keep this

File: test/test_optim.py [hunk 2]
```diff
@@ -2693,9 +2693,9 @@ def _diff_fn(p, grad, opt_differentiable_state, opt_class, kwargs, *ignored):
     p.grad = grad
     opt_differentiable_state = {k: v.clone() for k, v in opt_differentiable_state.items()}
     opt = opt_class([p], **kwargs)
-    opt.state.update(opt_differentiable_state)
+    opt.state[p].update(opt_differentiable_state)
```
Comment (emcastillo) [inline test/test_optim.py:2696]: This was my mistake in the SGD PR, state is a dict of dicts where the key is the param.

File: test/test_optim.py [hunk 3]
```diff
@@ -2693,9 +2693,9 @@ def _diff_fn(p, grad, opt_differentiable_state, opt_class, kwargs, *ignored):
...
     opt.step()
-    return (p,) + tuple(opt_differentiable_state.values())
+    return (p,) + tuple(v for v in opt_differentiable_state.values() if v.requires_grad)
```
Comment (albanD) [inline test/test_optim.py:2698]: Is this actually needed? Gradcheck will filter out things that don't require gradients already

File: test/test_optim.py [hunk 4]
```diff
@@ -2707,6 +2707,21 @@ def test_sgd(self):
         state = {'momentum_buffer': mbuff}
         gradcheck(_diff_fn, (p, grad, state, torch.optim.SGD, {'lr': 0.9, 'differentiable': True}, *state.values()))
 
+    def test_adam(self):
+        state = {}
+        p = torch.rand(10, requires_grad=True, dtype=torch.float64)
+        grad = torch.rand(10, requires_grad=True, dtype=torch.float64)
+        state['step'] = torch.tensor(10., requires_grad=False, dtype=torch.float64)
```
Comment (albanD) [inline test/test_optim.py:2714]: not requires grad?

File: torch/optim/adam.py (hunk 1 above)
Comment (albanD) [reply torch/optim/adam.py]: Ho right maximum saves both its inputs.

File: test/test_optim.py (hunk 4 above)
Comment (emcastillo) [reply test/test_optim.py:2714]: this causes the gradcheck to fail if its enabled ...

File: test/test_optim.py (hunk 3 above)
Comment (emcastillo) [reply test/test_optim.py:2698]: actually it didn't, lol step was returned and gradcheck detected errors with it

File: torch/optim/adam.py [hunk 5]
```diff
@@ -278,7 +280,7 @@ def _single_tensor_adam(params: List[Tensor],
 
             if amsgrad:
                 # Maintains the maximum of all 2nd moment running avg. till now
-                torch.maximum(max_exp_avg_sqs[i], exp_avg_sq, out=max_exp_avg_sqs[i])
+                max_exp_avg_sqs[i][:] = torch.maximum(max_exp_avg_sqs[i].clone(), exp_avg_sq)
```
Comment (albanD) [inline torch/optim/adam.py:283]: This is a very personal preference but I'm not very confident with advanced indexing assignment in general, can this be done with `copy_()` to make sure of the behavior we get?

But we should also avoid the clone in general.
```suggestion
                prev_exp_avg = max_exp_avg_sqs[i].clone() if differentiable else max_exp_avg_sqs[i]
                max_exp_avg_sqs[i].copy_(torch.maximum(prev_exp_avg, exp_avg_sq))
```

Comment (emcastillo) [reply torch/optim/adam.py:283]: the problem here is that if we pass the `differentiable` argument, the JIT and other pieces of code will complain and we will make this non-compat since torchscript does not allow arguments with default values.
I wonder if we can store the flag in the thread local storage set with the differentiable context manager for this check?

Comment (albanD) [reply torch/optim/adam.py:283]: It is only kwarg-only arguments with default values IIRC ?
You can make it positional as done in https://github.com/pytorch/pytorch/blob/1dfcad84aaad9c095b4dbbdafd5c351e9a712d45/torch/optim/sgd.py#L172-L175 to workaround the torchscript bug.

File: test/test_optim.py (hunk 4 above)
Comment (emcastillo) [reply test/test_optim.py:2714]: https://github.com/pytorch/pytorch/blob/master/torch/optim/adam.py#L140-L141
It was defined as float in the optimizer itself

Comment (albanD) [reply test/test_optim.py:2727]: Ok,
Can you leave a comment here that mentions that step is not a continuous variable (even though we define it as a float) and so it shouldn't require gradients.
btw can we have an nice assert in the `adam()` function to ensure that step never requires grad?

File: torch/optim/adam.py [hunk 6]
```diff
@@ -305,7 +315,7 @@ def _single_tensor_adam(params: List[Tensor],
 
             if amsgrad:
                 # Maintains the maximum of all 2nd moment running avg. till now
-                torch.maximum(max_exp_avg_sqs[i], exp_avg_sq, out=max_exp_avg_sqs[i])
+                torch.maximum(max_exp_avg_sqs[i].clone(), exp_avg_sq, out=max_exp_avg_sqs[i])
```
Comment (albanD) [inline torch/optim/adam.py:318]: Only if differentiable?

File: torch/optim/adam.py [hunk 7]
```diff
@@ -328,7 +338,7 @@ def _multi_tensor_adam(params: List[Tensor],
                        weight_decay: float,
                        eps: float,
                        maximize: bool,
-                       capturable: bool):
+                       differentiable: bool):
```
Comment (albanD) [inline torch/optim/adam.py:341]: The foreach ops don't support autograd. So we should have an assert here that differentiable is False (same for sgd btw).
This can be done in a separate PR though.

**Review by albanD — COMMENTED** [review]
nit on a clone but SGTM otherwise.
The CI needs fixing though no?

File: torch/optim/adam.py (hunk 6 above)
Comment (emcastillo) [reply torch/optim/adam.py:318]: this was not needed, the differentiable case is now handled in the outer if :)

File: torch/optim/adam.py (hunk 7 above)
Comment (emcastillo) [reply torch/optim/adam.py:343]: Added the assert to this one, will fix SGD too in the next PR :)

File: torch/optim/adam.py [hunk 8]
```diff
@@ -152,7 +154,8 @@ def step(self, closure=None):
 
                     if group['amsgrad']:
                         max_exp_avg_sqs.append(state['max_exp_avg_sq'])
-
+                    if group['differentiable']:
+                        assert not state['step'].requires_grad
```
Comment (albanD) [inline torch/optim/adam.py:158]: nit make this a RuntimeError and add a nice error message.

**Review by albanD — APPROVED** [review]
SGTM !
Just a small nit about error meessage

File: torch/optim/adam.py (hunk 8 above)
Comment (emcastillo) [reply torch/optim/adam.py]: done!
